src/client/src/producer.rs

```diff
@@ -62,21 +103,79 @@ impl TopicProducer {
-            "Connected to replica leader:"
-        );
-        let records: Vec<_> = records
+        let topics = self.pool.metadata.topics();
+        let topic_spec = topics
+            .lookup_by_key(&self.topic)
+            .await?
+            .ok_or_else(|| FluvioError::TopicNotFound(self.topic.to_string()))?
+            .spec;
+        let partition_count = topic_spec.partitions();
+        let partition_config = PartitionerConfig { partition_count };
+
+        let entries = records
             .into_iter()
-            .map(|(key, value): (Option<K>, V)| {
-                let key = key.map(|k| Bytes::from(k.into()));
-                let value = Bytes::from(value.into());
-                (key, value)
-            })
-            .collect();
-        send_records_raw(spu_client, &replica, records).await?;
+            .map::<(Option<Vec<u8>>, Vec<u8>), _>(|(k, v)| (k.map(|it| it.into()), v.into()))
+            .map(|(key, value)| {
+                let key = key.map(|it| DefaultAsyncBuffer::new(it));
+                let value = DefaultAsyncBuffer::new(value);
+                DefaultRecord::from((key, value))
+            });
+
+        // Calculate the partition for each entry
+        // Use a block scope to ensure we drop the partitioner lock
+        let records_by_partition = {
+            let mut iter = Vec::new();
+            let mut partitioner = self.partitioner.lock().await;
+            partitioner.update_config(partition_config);
+
+            for record in entries {
+                let key = record.key.as_ref().map(|k| k.as_ref());
+                let value = record.value.as_ref();
+                let partition = partitioner.partition(key, value);
+                iter.push((partition, record));
+            }
+            iter
+        };
+
+        // Group all of the records by the partitions they belong to, then
+        // group all of the partitions by the SpuId that leads that partition
+        let partitions_by_spu = group_by_spu(
+            &self.topic,
+            self.pool.metadata.partitions(),
+            records_by_partition,
+        )
+        .await?;
+
+        // Create one request per SPU leader
+        let mut requests: Vec<(SpuId, DefaultProduceRequest)> =
+            Vec::with_capacity(partitions_by_spu.len());
+        for (leader, partitions) in partitions_by_spu {
+            let mut request = DefaultProduceRequest::default();
+
+            let mut topic_request = DefaultTopicRequest::default();
+            topic_request.name = self.topic.clone();
+
+            for (partition, records) in partitions {
+                let mut partition_request = DefaultPartitionRequest::default();
+                partition_request.partition_index = partition;
+                partition_request
+                    .records
+                    .batches
+                    .push(DefaultBatch::new(records));
+                topic_request.partitions.push(partition_request);
```

Comment:
unit test?

Reply:
Factored out `assemble_requests` method and wrote a test for it